### PR TITLE
feat: App index updating

### LINKED ISSUE #73: implement app updates for the app index
* submitted apps should have an `app_id` and `version` fields in the manifest.toml file
* the bot will detect if the imported app is an upgrade of an already existing app on the index
* when the xdcstore frontend request the bot to send the updated index metadata, it also sends the app_id and version of downloaded apps
* when the bot receives an update request from the xdcstore frontend, besides the app index metadata it also checks the `app_id`s and `version`s of downladed apps and if there is a newer version, sends them one by one in new status-updates to the xdcstore frontend.
* use integer for `app_id`

hence update of downloaded apps happen automatically when the user use the app and a "update index" request is sent, because of this, a way to remove downloaded apps from cache/local-storage is needed, so user have a way to stop the updates of an app downloaded once to try but never wanted/used anymore, for that aspect see issue https://github.com/webxdc/store/issues/123

### PR DESCRIPTION
**Pull request by Septias**
Implement app index updating. 

Importing with CLi will scan the import folder and check every xdc with `app_id` and `version` to determine the importing state:
- Added (New app is added because app_id doesn't exist)
- Updated (Old AppInfo is invalidated and a new one is created with a new xdc file)
- Ignored (`version` and `app_info` already exist in this combination so don't do anything)

closes #73

### CODE REVIEW

File: src/db.rs [hunk 1]
```diff
@@ -373,9 +374,10 @@ pub async fn create_app_info(
         .bind(&app_info.source_code_url)
         .bind(next_serial)
         .bind(app_info.id)
-        .execute(c)
+        .execute(&mut *trans)
         .await?;
     app_info.id = i32::try_from(res.last_insert_rowid())?;
```
Comment (Septias) [inline src/db.rs:379]: This is somewhat unrelated but should happen anyway

File: src/request_handlers/shop.rs [hunk 2]
```diff
@@ -27,7 +27,7 @@ use ts_rs::TS;
 enum ShopRequest {
     Update {
         /// Requested update sequence number.
-        serial: i32,
+        serial: u32,
```
Comment (Septias) [inline src/request_handlers/shop.rs:30]: Unrelated, but serial is stored as u32 in the db anyway.

**Review by adbenitez — DISMISSED** [review]

File: src/db.rs [hunk 3]
```diff
@@ -693,4 +738,76 @@ mod tests {
...
+
+    #[tokio::test]
+    async fn upgrade_app() {
+        let mut conn = SqliteConnection::connect("sqlite::memory:").await.unwrap();
+        MIGRATOR.run(&mut conn).await.unwrap();
+        set_config(&mut conn, &BotConfig::default()).await.unwrap();
+
+        let mut app_info = AppInfo {
+            app_id: "testxdc".to_string(),
+            version: "1.0.0".to_string(),
```
Comment (hpk42) [inline src/db.rs:750]: i guess you rather want to migraate to version numbers in a separate PR, right?

Comment (Septias) [reply src/db.rs:750]: yes

File: frontend/src/db/shop_db.ts [hunk 4]
```diff
@@ -69,6 +69,17 @@ export class AppInfoDB {
     })
   }
 
+  async remove_multiple(ids: number[]): Promise<void> {
```
Comment (hpk42) [inline frontend/src/db/shop_db.ts:72]: please use more explicit naming like
```suggestion
  async remove_multiple_apps(app_ids: number[]): Promise<void> {
```

File: src/db.rs [hunk 5]
```diff
@@ -444,6 +447,46 @@ pub async fn get_active_app_infos_since(
...
+    serial: u32,
+) -> sqlx::Result<Vec<AppInfo>> {
+    sqlx::query_as::<_, DBAppInfo>("SELECT * FROM app_infos WHERE active = false AND serial > ?")
+        .bind(serial)
+        .fetch_all(c)
+        .await
+        .map(|app| app.into_iter().map(|a| a.into()).collect())
+}
+
+pub async fn app_info_exists(c: &mut SqliteConnection, app_id: &str) -> sqlx::Result<bool> {
```
Comment (hpk42) [inline src/db.rs:461]:
```suggestion
pub async fn app_exists(c: &mut SqliteConnection, app_id: &str) -> sqlx::Result<bool> {
```

File: src/main.rs [hunk 6]
```diff
@@ -80,26 +84,46 @@ async fn main() -> anyhow::Result<()> {
...
                             fs::copy(file, &new_path).with_context(|| {
                                 format!(
-                                    "failed to copy {} to {}",
+                                    "Failed to copy {} to {}",
                                     file.display(),
                                     new_path.display()
                                 )
                             })?;
 
                             app_info.xdc_blob_dir = new_path;
```
Comment (hpk42) [inline src/main.rs:93]: why is this called "dir" here btw?  Isn't it the path to the app `.xdc` file?

File: src/request_handlers/shop.rs [hunk 7]
```diff
@@ -51,7 +51,10 @@ pub enum ShopResponse {
         error: String,
     },
     Update {
+        /// List of new app ids.
         app_infos: Vec<AppInfo>,
+        /// List of removed app ids.
```
Comment (hpk42) [inline src/request_handlers/shop.rs:56]: please document more verbosely: removed versions or removed all-versions of an app_id?

File: src/utils.rs [hunk 8]
```diff
@@ -69,15 +69,16 @@ pub async fn send_newest_updates(
...
-        .into_iter()
-        .collect();
-
+    let app_infos: Vec<_> = db::get_active_app_infos_since(db, serial).await?;
+    let removed = db::get_inactive_app_infos_since(db, serial).await?;
     let serial = db::get_last_serial(db).await?;
-    let resp = ShopResponse::Update { app_infos, serial };
+    let resp = ShopResponse::Update {
+        app_infos,
+        removed: removed.into_iter().map(|app_info| app_info.id).collect(),
```
Comment (hpk42) [inline src/utils.rs:79]: what is this "removed" list of app_id's meant to signal to the store.xdc? 
i first thought that "removed" means that the whole app (with all versions) got removed but `get_inactive_app_infos_since` provides all old versions.  isn't "removed"  non-empty for a simply upgraded app (i.e. there was version 1, now there is version 2)?

**Review by hpk42 — CHANGES_REQUESTED** [review]
see the reviews, questions i left -- 
please clarify what the code around "removing" is meant to achieve. 
I also don't get why old versions need to be marked "inactive" -- why not simply store new versions and later determine what is the newest ("active") one when sending the update-struct to the frontend?  It shouldn't need an extra column unless i am missing something.

File: src/main.rs (hunk 6 above)
Comment (Septias) [reply src/main.rs:93]: yes, should be path

File: src/request_handlers/shop.rs (hunk 7 above)
Comment (Septias) [reply src/request_handlers/shop.rs:56]: Sorry, this should be `Removed appinfos` or so. It is a list of ids (row-ids to be explicit) which are no longer active. (active = false)

File: src/utils.rs (hunk 8 above)
Comment (Septias) [reply src/utils.rs:79]: Yes it not empty as it removes the old appinfo and adds a new one (that for version 2)

File: src/main.rs (hunk 6 above)
Comment (Septias) [reply src/main.rs:93]: But that's also a refactoring I would like to do outside this pr

File: src/db.rs (hunk 5 above)
Comment (Septias) [reply src/db.rs:461]: This is supposed to be app_info actually.

**Comment by Septias** [comment]
We can refactor this `active` column usage to only use the most recent version if that is preferred. When I implemented it was still thinking about sub-version updates to apps (small fixes) that would not need a version change. But we can also for the simple approach of just thinking `some changes` -> `increased version`

**Comment by Septias** [comment]
Closing this for now as we discussed a new approach. Will create a follow-up soon.

**Comment by hpk42** [comment]
just discussed a bit IRL with @septias -- so plan is to remove "removed" list and also the "inactive" column -- likely sebi is going to attempt all of this tomorrow.

**Pull request closed without merging**